posit-dev/positron
# DRAFT: Two-way communication for IPyWidgets

### Intent

This (draft) PR implements two-way communication inside a webview for ipywidgets. This is very similar to functionality implemented in the vscode-jupyter extension but here it's using our runtimes (I was pleasantly surprised that our `performRpc` _just works_ with ipywidgets!). It's still pretty rough, but I wanted to get feedback about the approach before polishing.

### Approach

The approach is to pass messages to and from the webview as needed to implement two interfaces required by the jupyter-widgets libraries: `IClassicComm` and `IWidgetManager`. I implemented the minimum amount required for both by trial-and-error while walking through the [IPyWidgets introduction](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20Basics.html).

I roughly followed some practices used in the `markdown-language-features` and `simple-browser` extensions, where they also bundle the webview code (in the preview-src subdirectory) using esbuild. I packaged it in an extension (without any extension code yet) for now.

Lots of things to be reconsidered/improved, but I think this gets us most of the way there. I'll be on leave over the next two weeks but wanted to at least share this as a draft before I go.

Known issues:

- It doesn't clean up comms and seems to create duplicates when the webview loses and gains focus.
- There's a lot of code in jupyter-widgets around a `buffer` concept that I've not implemented. I'm not sure what that does.
- I added `metadata` to our Jupyter messages since it's needed by the ipywidgets Python lib. We could alternatively try to customize the Python lib if we don't want to add that arg.
- It's all just dumped into the `PositronNotebookOutputWebviewService.createWidgetHtmlOutput` method and could be better organized.
- Probably more!

### QA Notes

To test this, you _should_ be able to checkout the branch then do:

```sh
cd extensions/positron-ipywidgets
yarn run build-webview
```

(I haven't figured out the right way to hook that into the existing build and compile/watch tasks.)

Then start Positron and run code that creates and displays a widget:

```python
import ipywidgets as widgets
a = widgets.FloatSlider()
display(a)
```

This should show a slider in the plots pane. Adjust the slider, then check that it updated the value in the kernel:

```python
a.value
# should match the value in the UI
```

You can also update the value in the kernel, and it should update to match that in the UI:

```python
a.value = 100
# should move the slider in the UI
```

We should also walk through more of the [IPyWidgets docs](https://ipywidgets.readthedocs.io/en/stable/examples/Widget%20Basics.html) to more thoroughly test this.

Here's a quick demo of the above:

https://github.com/posit-dev/positron/assets/559360/d0b1fbdc-7489-4808-b8f4-456427ecc6b6
